Doxygen/man/man3/sql.3
.TH "sql" 3 "Fri Feb 3 2017" "Version Version: alpha v1.5" "Git Gud: The adventures of the hungover" \" -*- nroff -*-
.ad l
.nh
.SH NAME
sql
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <sql\&.hpp>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBsql\fP ()"
.br
.ti -1c
.RI "\fB~sql\fP ()"
.br
.ti -1c
.RI "std::vector< std::string > \fBget_data\fP (char *table, char *columnname)"
.br
.RI "Get data from a column in a table in the database\&. "
.ti -1c
.RI "std::vector< std::string > \fBget_data\fP (char *table, char *columnname, char *condition1, char *condition2)"
.br
.RI "Get data from a column in a table in the database with condition\&. "
.ti -1c
.RI "void \fBchange_data\fP (char *table, char *condition1, char *condition2, char *columnname, char *new_value)"
.br
.RI "Change values in the database\&. "
.ti -1c
.RI "void \fBadd_data\fP (char *table, const char *values)"
.br
.RI "Add a new value to database\&. "
.ti -1c
.RI "void \fBexecute_query_without_return\fP (const char *to_be_executed)"
.br
.RI "Execute a query\&. "
.ti -1c
.RI "std::vector< std::string > \fBexecute_query_with_return\fP (const char *to_be_executed)"
.br
.RI "Execute a query\&. "
.ti -1c
.RI "std::vector< std::string > \fBget_name_from_id_inventory\fP (std::string item_id)"
.br
.RI "Get the name from an item\&. "
.ti -1c
.RI "std::vector< std::string > \fBget_sprite_inventory\fP (std::string item_id)"
.br
.RI "Get the sprite_path from an item\&. "
.ti -1c
.RI "std::vector< std::string > \fBget_in_game_name_from_id_inventory\fP (std::string item_id)"
.br
.RI "Get the name from an item\&. "
.ti -1c
.RI "std::vector< std::string > \fBget_inventory\fP ()"
.br
.RI "Get the inventory\&. "
.ti -1c
.RI "std::vector< std::string > \fBget_player_value\fP (int save, std::string variable_name)"
.br
.RI "Get value of a column from the player\&. "
.ti -1c
.RI "std::vector< std::string > \fBget_player_sprite\fP (int save)"
.br
.RI "Get sprite sheet of the player\&. "
.ti -1c
.RI "std::vector< std::string > \fBget_player_value\fP (std::string npc_id)"
.br
.RI "Get value of the player\&. "
.ti -1c
.RI "void \fBadd_item_to_inventory\fP (std::string item_id)"
.br
.RI "Add an item to the players inventory\&. "
.ti -1c
.RI "std::vector< std::string > \fBget_npc_value\fP (std::string npc_id)"
.br
.RI "Get all values from columns in NPC\&. "
.ti -1c
.RI "std::vector< std::string > \fBget_number_of_npcs\fP ()"
.br
.RI "Get all id's of the NPC's\&. "
.ti -1c
.RI "std::vector< std::string > \fBget_object_sprite_ids\fP ()"
.br
.RI "Get id's from objects and tiles sprites\&. "
.ti -1c
.RI "std::vector< std::string > \fBget_object_sprite_value\fP (std::string object_id)"
.br
.RI "Get all values from columns in sprites\&. "
.ti -1c
.RI "void \fBset_new_map\fP (std::string map_name, int background_music)"
.br
.RI "Create new map\&. "
.ti -1c
.RI "void \fBsave_tiles\fP (std::string tile_path, int pos_x, int pos_y, std::string map_name)"
.br
.RI "Saves the tiles in the database\&. "
.ti -1c
.RI "std::vector< std::string > \fBget_available_levels\fP ()"
.br
.RI "Get the names of all levels in the database\&. "
.ti -1c
.RI "std::vector< std::string > \fBget_level_ids\fP ()"
.br
.RI "Get the ID's of all levels in the database\&. "
.ti -1c
.RI "std::map< std::string, std::vector< std::string > > \fBget_level_object_value\fP (std::string level_id)"
.br
.RI "Get all object values of a level\&. "
.ti -1c
.RI "sf::IntRect \fBget_collision_objects\fP (std::vector< std::string > objects)"
.br
.RI "Get the not walkable objects\&. "
.ti -1c
.RI "std::map< std::string, std::vector< std::string > > \fBget_level_background_value\fP (std::string level_id)"
.br
.RI "Get all background values of a level\&. "
.ti -1c
.RI "std::vector< std::string > \fBget_music\fP (std::string map_id)"
.br
.RI "Get music of consisting map\&. "
.ti -1c
.RI "std::vector< std::string > \fBget_quest_text\fP (std::string quest_id, std::string part)"
.br
.RI "Get quest part text\&. "
.ti -1c
.RI "std::vector< std::string > \fBget_quest_parts\fP (std::string quest_id)"
.br
.RI "Get all part numbers from quest\&. "
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static std::vector< std::string > \fBget_result\fP ()"
.br
.RI "Get saved result\&. "
.ti -1c
.RI "static void \fBset_result\fP (char *\fBx\fP)"
.br
.RI "Save result to static value\&. "
.ti -1c
.RI "static int \fBcallback\fP (void *NotUsed, int argc, char **argv, char **azColName)"
.br
.RI "Get value from database\&. "
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "void \fBcheck_query_request\fP ()"
.br
.RI "Checks if query is executed correctly\&. "
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "sqlite3 * \fBdatabase\fP"
.br
.ti -1c
.RI "std::string \fBdatabase_path\fP = '\&.\&./\&.\&./bin/database\&.db'"
.br
.ti -1c
.RI "int \fBrc\fP"
.br
.ti -1c
.RI "int \fBconnection\fP"
.br
.ti -1c
.RI "char * \fB_ERROR_MESSAGE\fP = 0"
.br
.ti -1c
.RI "const char * \fBdata\fP = 'Callback function called'"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 13 of file sql\&.hpp\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "sql::sql ()"

.PP
Definition at line 6 of file sql\&.cpp\&.
.SS "sql::~sql ()"

.PP
Definition at line 24 of file sql\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "void sql::add_data (char * table, const char * values)"

.PP
Add a new value to database\&. This method adds a new value in the given table of the database\&.
.br
The parameter 'value' is consists all the column values except the id,
.br
because the this value has auto increment\&.
.PP
.SS "Parameters: "
.PP
table : char* 
.br
-> Table name in the database
.PP
values : char* 
.br
-> column values to add in the database\&.
.PP
.SS "Example: "
.PP
add_data('sprites',''\&.\&./\&.\&./bin/objects/new_pic\&.png', new pic , 64 , 64')
.br
There is now a new record in the database with the given values\&. 
.PP
Definition at line 74 of file sql\&.cpp\&.
.SS "void sql::add_item_to_inventory (std::string item_id)"

.PP
Add an item to the players inventory\&. This function will give the player an item that will be stored in the inventory table in the database\&.
.br
.PP
.SS "Parameters: "
.PP
item_name : std::string -> The name of the item that has to be given
.PP
.SS "Example: "
.PP
add_item_to_inventory('1'); -> Adds a banana to the users inventory 
.PP
Definition at line 83 of file sql\&.cpp\&.
.SS "int sql::callback (void * NotUsed, int argc, char ** argv, char ** azColName)\fC [static]\fP"

.PP
Get value from database\&. This method is used to get the value from the database\&. What the function does is explained in the sqlite3\&.hpp file\&. In this class we use it to store the found data in a global static value\&. 
.PP
Definition at line 29 of file sql\&.cpp\&.
.SS "void sql::change_data (char * table, char * condition1, char * condition2, char * columnname, char * new_value)"

.PP
Change values in the database\&. This method overwrites data in the database, where the value of the given column meets the condition values\&.
.PP
.SS "Parameters: "
.PP
table : char* 
.br
-> Table name in the database
.PP
condition1 : char* 
.br
-> First condition of the Where statement in the SQL query\&.
.PP
condition2 : char* 
.br
-> Second condition of the Where statement in the SQL query\&.
.PP
columnname : char* 
.br
-> Column name in the database
.PP
new_value : char* 
.br
-> New value to be set in the specified column value
.PP
.SS "Example: "
.PP
change_data('player','id','1','name','pieter')
.br
The name of the player is now 'pieter' instead of 'arno' 
.PP
Definition at line 93 of file sql\&.cpp\&.
.SS "void sql::check_query_request ()\fC [private]\fP"

.PP
Checks if query is executed correctly\&. This method checks if the executed query is correctly procesed\&.
.br
If that is not the case this function will print an error on the command prompt\&.
.br
.PP
.SS "Example: "
.PP
\fBcheck_query_request()\fP
.br
\fIno error, so do nothing\fP 
.PP
Definition at line 36 of file sql\&.cpp\&.
.SS "std::vector< std::string > sql::execute_query_with_return (const char * to_be_executed)"

.PP
Execute a query\&. This method executes a given query that has a return value\&.
.br
 
.SS "Parameters: "
.PP
to_be_executed : const char* 
.br
-> Query to be executed
.PP
.SS "Return: "
.PP
list_of_data : std::vector<std::string> 
.br
-> List that contains the asked data in the form of strings\&.
.PP
.SS "Example: "
.PP
execute_query_with_return('SELECT name FROM player WHERE id == 1;')
.br
Return: {'arno'} 
.PP
Definition at line 48 of file sql\&.cpp\&.
.SS "void sql::execute_query_without_return (const char * to_be_executed)"

.PP
Execute a query\&. This method executes a given query that has no return value\&.
.br
 
.SS "Parameters: "
.PP
to_be_executed : const char* 
.br
-> Query to be executed
.PP
.SS "Example: "
.PP
execute_query_without_return('INSERT INTO sprites VALUES(NULL, '\&.\&./\&.\&./bin/objects/new_pic\&.png', new pic , 64 , 64);')
.br
There is now a new record in the database with the given values\&. 
.PP
Definition at line 43 of file sql\&.cpp\&.
.SS "std::vector< std::string > sql::get_available_levels ()"

.PP
Get the names of all levels in the database\&. This method returns all names of the levels that are in the database\&.
.br
This can be usefull so you can know if a level already exists\&. 
.br
.PP
.SS "Return: "
.PP
list_of_data : std::vector<std::string> 
.br
-> List that contains the asked data in the form of strings\&.
.PP
.SS "Example: "
.PP
\fBget_available_levels()\fP
.br
-> {'level1','level2','level5'} 
.PP
Definition at line 389 of file sql\&.cpp\&.
.SS "sf::IntRect sql::get_collision_objects (std::vector< std::string > objects)"

.PP
Get the not walkable objects\&. This method returns the hitbox of an object that is not walkable\&.
.br
The hitbox can be used to calculate the collision detection on\&. 
.br
 
.SS "Parameters: "
.PP
objects : std::vector<std::string> 
.br
-> values of one opbject\&.
.PP
.SS "Return: "
.PP
collision_rects : sf::IntRect 
.br
-> Hitbox of the not walkable object\&.
.br
.PP
.SS "Example: "
.PP
get_collision_objects({\&.\&.\&.}) 
.br
-> (100,100,50,50) 
.PP
Definition at line 502 of file sql\&.cpp\&.
.SS "std::vector< std::string > sql::get_data (char * table, char * columnname)"

.PP
Get data from a column in a table in the database\&. This method gets the data from a column in a table in the database\&.
.br
 
.SS "Parameters: "
.PP
table : char* 
.br
-> Table name in the database
.PP
column : char* 
.br
-> Column name in the database
.PP
.SS "Return: "
.PP
list_of_data : std::vector<std::string> 
.br
-> List that contains the asked data in the form of strings\&.
.PP
.SS "Example: "
.PP
Get_data('player', 'level') 
.br
return: {'1'}\&. 
.PP
Definition at line 66 of file sql\&.cpp\&.
.SS "std::vector< std::string > sql::get_data (char * table, char * columnname, char * condition1, char * condition2)"

.PP
Get data from a column in a table in the database with condition\&. This method gets the data from a table in the database,
.br
where the value of the given column meets the condition values\&.
.PP
.SS "Parameters: "
.PP
table : char* 
.br
-> Table name in the database
.PP
column : char* 
.br
-> Column name in the database
.PP
condition1 : char* 
.br
-> First condition of the Where statement in the SQL query\&.
.PP
condition2 : char* 
.br
-> Second condition of the Where statement in the SQL query\&.
.PP
.SS "Return: "
.PP
list_of_data : std::vector<std::string> 
.br
-> List that contains the asked data in the form of strings\&.
.PP
.SS "Example: "
.PP
Get_data('player', 'name','id','1') 
.br
return: {'arno'}\&. 
.PP
Definition at line 57 of file sql\&.cpp\&.
.SS "std::vector< std::string > sql::get_in_game_name_from_id_inventory (std::string item_id)"

.PP
Get the name from an item\&. This function returns the name of an item by using its ID\&.
.br
 
.SS "Parameters: "
.PP
item_id : std::string 
.br
-> The ID of the item\&.
.PP
.SS "Return: "
.PP
name : std::vector<std::string> 
.br
-> The in game name of the requested item\&.
.PP
.SS "Example: "
.PP
get_name_from_id_inventory('1') 
.br
-> Tokato Gloves 
.PP
Definition at line 144 of file sql\&.cpp\&.
.SS "std::vector< std::string > sql::get_inventory ()"

.PP
Get the inventory\&. This function retrieves the inventory and its values and puts it into a vector\&. 
.br
.PP
.SS "Return: "
.PP
inventory_vector : std::vector<std::string> 
.br
-> This vector contains the items in the inventory
.PP
.SS "Example: "
.PP
\fBget_inventory()\fP 
.br
-> {'1', '2', '2', '3'} 
.PP
Definition at line 114 of file sql\&.cpp\&.
.SS "std::map< std::string, std::vector< std::string > > sql::get_level_background_value (std::string level_id)"

.PP
Get all background values of a level\&. This method returns the values of all background tiles that are assigned to the given level ID\&.
.br
.PP
.SS "Parameters: "
.PP
level_id : std::string 
.br
-> id of the level from what the background tiles have to be loaded\&.
.PP
.SS "Return: "
.PP
list_of_data : std::map< std::string, std::vector<std::string> > 
.br
-> Map that contains the asked data in the form of strings with name as key\&.
.br
.PP
.SS "Example: "
.PP
get_level_object_value('level1') 
.br
{'road1':{'\&.\&./\&.\&./bin/tiles/example\&.png','120','150','1'}} 
.PP
Definition at line 452 of file sql\&.cpp\&.
.SS "std::vector< std::string > sql::get_level_ids ()"

.PP
Get the ID's of all levels in the database\&. This method returns all ID's of the levels that are in the database\&.
.br
.PP
.SS "Return: "
.PP
list_of_data : std::vector<std::string> 
.br
-> List that contains the asked data in the form of strings\&.
.PP
.SS "Example: "
.PP
\fBget_level_ids()\fP
.br
-> {'1','2','5'} 
.PP

Software/Maingame/src/sql.cpp
#include "sql.hpp"




sql::sql() {
	
	//Checking if the given database exists and not locked.
	std::ifstream database_existance(database_path);
	if (!database_existance.good()) {
		std::cout	<< "Error: Could not open database."				//error definition
					<< "Probbably incorrect path or locked file.\n";	//error explanation
		while (1);
	}
	else { 
		std::cout << "Database correctly opened.\n";					//database found
	}
	database_existance.close();

	//Opening the database
	connection = sqlite3_open(database_path.c_str(), &database);
}	

sql::~sql() {
	sqlite3_close(database);
}


int sql::callback(void *NotUsed, int argc, char **argv, char **azColName) {
	for (int i = 0; i < argc; i++) {
		set_result(argv[0]);
	}
	return 0;
}

void sql::check_query_request() {
	if (rc != SQLITE_OK) {
		std::cout << "ERROR: " << _ERROR_MESSAGE << "\tLine: " << __LINE__ << std::endl;
		sqlite3_free(_ERROR_MESSAGE);
	}
}

void sql::execute_query_without_return(const char* to_be_executed) {
	while (rc = sqlite3_exec(database, to_be_executed, callback, 0, &_ERROR_MESSAGE) != 0);
	check_query_request();
}

std::vector<std::string> sql::execute_query_with_return(const char* to_be_executed) {

	while (rc = sqlite3_exec(database, to_be_executed, callback, 0, &_ERROR_MESSAGE) != 0);
	check_query_request();

	return get_result();
}


std::vector<std::string> sql::get_data(char* table, char* columnname, char* condition1, char* condition2) {
	std::ostringstream query;
	query << "SELECT " << columnname << " FROM " << table << " WHERE " << condition1 << " == " << condition2 << ";";
	
	std::string query_string = query.str();
	return execute_query_with_return(query_string.c_str());
}


std::vector<std::string> sql::get_data(char* table, char* columnname) {
	std::ostringstream query;
	query << "SELECT " << columnname << " FROM " << table << ";";
	
	std::string query_string = query.str();
	return execute_query_with_return(query_string.c_str());
}

void sql::add_data(char* table, const char* values) {
	std::ostringstream query;
	query << "INSERT INTO " << table << " VALUES( NULL, " << values << ");";

	std::string query_string = query.str();
	execute_query_without_return(query_string.c_str());
}


void sql::add_item_to_inventory(std::string item_id) {
	std::ostringstream query;
	query << "INSERT INTO inventory VALUES( " << item_id << ", 1, 0);";

	std::string query_string = query.str();
	std::cout << "Going to update" << std::endl;
	execute_query_without_return(query_string.c_str());
	std::cout << "Updated" << std::endl;
}

void sql::change_data(char* table, char* condition1, char* condition2, char* columnname, char* new_value) {
	std::ostringstream query;
	query	<< "UPDATE "	<< table 
			<< " SET "		<< columnname << " = '"	<< new_value
			<< "' WHERE "	<< condition1	<< " = "	<< condition2 << ";";
	
	std::string query_string = query.str();
	execute_query_without_return(query_string.c_str());
}

std::vector<std::string> sql::get_result() {
	std::vector<std::string>temp_result_save;
	temp_result_save = result;
	while (result.size() > 0) {
		result.pop_back();
	}
	return temp_result_save;

}


std::vector<std::string> sql::get_inventory() {
	std::ostringstream query;
	query << "SELECT item_id "
		<< " FROM inventory;";

	std::string query_string = query.str();
	return execute_query_with_return(query_string.c_str());

}

std::vector<std::string> sql::get_name_from_id_inventory(std::string item_id) {
	std::ostringstream query;
	query << "SELECT name "
		<< " FROM item"
		<< " WHERE id = " << item_id << ";";

	std::string query_string = query.str();
	return execute_query_with_return(query_string.c_str());
}

std::vector<std::string> sql::get_sprite_inventory(std::string item_id) {
	std::ostringstream query;
	query << "SELECT sprite "
		<< " FROM item"
		<< " WHERE id = " << item_id << ";";

	std::string query_string = query.str();
	return execute_query_with_return(query_string.c_str());
}

std::vector<std::string> sql::get_in_game_name_from_id_inventory(std::string item_id) {
	std::ostringstream query;
	query << "SELECT in_game_name "
		<< " FROM item"
		<< " WHERE id = " << item_id << ";";

	std::string query_string = query.str();
	return execute_query_with_return(query_string.c_str());
}


std::vector<std::string> sql::get_player_value(int save, std::string variable_name) {
	std::ostringstream query;
	query	<< "SELECT " << variable_name
			<< " FROM player "
			<< "WHERE id == " << save << ";";

	std::string query_string = query.str();
	return execute_query_with_return(query_string.c_str());
}


std::vector<std::string> sql::get_player_sprite(int save) {
	std::ostringstream query;
	query	<< "SELECT sprites.path "
			<< "FROM sprites, player "
			<< "WHERE sprites.id = player.sprite_id "
			<< "AND name = " << save << ";";

	std::string query_string = query.str();
	return execute_query_with_return(query_string.c_str());
}


std::vector<std::string> sql::get_number_of_npcs() {
	std::vector<std::string>total = get_data("NPC", "id");
	return total;
}

std::vector<std::string> sql::get_npc_value(std::string npc_id) {

	std::vector<std::string>temp_npc_value;
	std::vector<std::string>npc_values;
	std::string temp_query_string;
	std::ostringstream ostring_query;

	ostring_query 
	<< "SELECT "<< "name"			
	<< " FROM "	<< "NPC"		
	<< " WHERE "<< "id"	<< " == " << npc_id << ";"
	
	<< "SELECT "<< "sprites.path"	
	<< " FROM "	<< "sprites, NPC"
	<< " WHERE "<< "sprites.name"	<< " == " << "NPC.name" 
	<< " AND "	<< "NPC.id"		<< " == " << npc_id << ";"
	
	<< "SELECT " << "class_id"		
	<< " FROM "	 << "NPC"		
	<< " WHERE " << "id" << " == " << npc_id << ";"

	<< "SELECT " << "position_x"	
	<< " FROM "  << "NPC"		
	<< " WHERE " << "id" << " == " << npc_id << ";"
	
	<< "SELECT " << "position_y"	
	<< " FROM "  << "NPC"		
	<< " WHERE " << "id" << " == " << npc_id << ";"
	
	<< "SELECT " << "quest_id"		
	<< " FROM "  << "NPC"		
	<< " WHERE " << "id" << " == " << npc_id << ";";

	temp_query_string = ostring_query.str();
	temp_npc_value = execute_query_with_return(temp_query_string.c_str());
	for (auto indexer : temp_npc_value) {
		npc_values.push_back(indexer);
	}
	
	return npc_values;
}


std::vector<std::string> sql::get_object_sprite_ids() {
	
	std::vector<std::string>sprite_values;
	std::string temp_query_string;
	std::ostringstream ostring_query;
	
	ostring_query << "SELECT id FROM sprites WHERE path LIKE '../../bin/tiles%' OR path LIKE '../../bin/objects%';";
	temp_query_string = ostring_query.str();
	
	return execute_query_with_return(temp_query_string.c_str());
}


std::vector<std::string> sql::get_object_sprite_value(std::string object_id) {

	std::vector<std::string>temp_sprite_values;
	std::vector<std::string>sprite_values;
	std::string temp_query_string;
	std::ostringstream ostring_query;
	
	ostring_query
		<< "SELECT " << "path"
		<< " FROM " << "sprites"
		<< " WHERE " << "id" << " == " << object_id << ";"

		<< "SELECT " << "name"
		<< " FROM " << "sprites"
		<< " WHERE " << "id" << " == " << object_id << ";"

		<< "SELECT " << "size_x"
		<< " FROM " << "sprites"
		<< " WHERE " << "id" << " == " << object_id << ";"

		<< "SELECT " << "size_y"
		<< " FROM " << "sprites"
		<< " WHERE " << "id" << " == " << object_id << ";"

		<< "SELECT " << "walkable"
		<< " FROM " << "sprites"
		<< " WHERE " << "id" << " == " << object_id << ";";


	temp_query_string = ostring_query.str();
	temp_sprite_values = execute_query_with_return(temp_query_string.c_str());
	for (auto indexer : temp_sprite_values) {
		sprite_values.push_back(indexer);
	}

	return sprite_values;
}



void sql::set_new_map(std::string map_name, int background_music){
	std::ostringstream ostring_query;

	ostring_query 
		<< "INSERT INTO maps "
		<< "VALUES( " 
			<< "NULL,"
			<<" '" << map_name << "'," 
			<< std::to_string(background_music) 
		<< ");";
	execute_query_without_return(ostring_query.str().c_str());
}


void sql::save_tiles(std::string tile_path, int pos_x, int pos_y, std::string map_name){
	//checks
	std::vector<std::string>tile_paths = get_data("sprites", "path");
	std::vector<std::string>map_names = get_data("maps", "name");
	bool incorrect_path = false;
	bool incorrect_map = false;
	for (auto path : tile_paths) { if (path == tile_path) { incorrect_path = true; } }
	for (auto name : map_names) { if (name == map_name) { incorrect_map = true; } }



	if (incorrect_path && incorrect_map) {
		//saving
		std::vector<std::string>searched_values;
		
		std::string temp_query_string;

		std::size_t found = tile_path.find("/objects/");

		std::ostringstream sprite_name_from_path_query;
		std::ostringstream sprite_id_from_path_query;
		std::ostringstream map_id_from_map_name_query;
		std::ostringstream ostring_query;


		sprite_name_from_path_query
			<< "SELECT " << "name"
			<< " FROM " << "sprites"
			<< " WHERE " << "path" << " == '" << tile_path << "';";

		sprite_id_from_path_query
			<< "SELECT " << "id"
			<< " FROM " << "sprites"
			<< " WHERE " << "path" << " == '" << tile_path << "';";


		map_id_from_map_name_query
			<< "SELECT " << "id"
			<< " FROM " << "maps"
			<< " WHERE " << "name" << " == '" << map_name << "';";

		std::string temp_sprite_name_from_path_query = sprite_name_from_path_query.str();
		std::string temp_sprite_id_from_path_query = sprite_id_from_path_query.str();
		std::string temp_map_id_from_map_name = map_id_from_map_name_query.str();

		searched_values.push_back(execute_query_with_return(temp_sprite_name_from_path_query.c_str()).at(0));
		searched_values.push_back(execute_query_with_return(temp_sprite_id_from_path_query.c_str()).at(0));
		searched_values.push_back(execute_query_with_return(temp_map_id_from_map_name.c_str()).at(0));

		
		if (found != std::string::npos) {
			std::cout << "Word opgeslagen in objects.\n";
			ostring_query
				<< "INSERT INTO objects "
				<< "VALUES( "
				<< "NULL," << " '"
				<< searched_values.at(0) << "', "
				<< searched_values.at(1) << ", "
				<< std::to_string(pos_x) << ", "
				<< std::to_string(pos_y) << ", "
				<< searched_values.at(2)
				<< ");";
			execute_query_without_return(ostring_query.str().c_str());
		}
		else if ((found = tile_path.find("/tiles/")) != std::string::npos) {
			std::cout << "Word opgeslagen in background.\n";
			ostring_query
				<< "INSERT INTO background "
				<< "VALUES( "
				<< "NULL," << " '"
				<< searched_values.at(0) << "', "
				<< searched_values.at(1) << ", "
				<< std::to_string(pos_x) << ", "
				<< std::to_string(pos_y) << ", "
				<< searched_values.at(2)
				<< ");";
			execute_query_without_return(ostring_query.str().c_str());
		}
		else {
			std::cout << "Error: Tile not found.\t" //declaration
				<< "The map can not be saved due an incorrect or non-existing path."; //definition
		}
	}
	
	
	if(!incorrect_path){
		std::cout << "Error: Incorrect path.\t" //definition
			<< "The map can not be saved due an incorrect or non-existing tile path."; //declaration
	}
	if (!incorrect_map) {
		std::cout << "Error: Incorrect map name.\t"	//definition
			<< "The map can not be saved due an incorrect or non-existing level name."; //declaration
	}
}


std::vector<std::string> sql::get_available_levels() {
	std::vector<std::string>total = get_data("maps", "name");
	return total;
}

std::vector<std::string> sql::get_level_ids() {
	std::vector<std::string>total = get_data("maps", "id");
	return total;
}



std::map<std::string, std::vector<std::string>> sql::get_level_object_value(std::string level_id) {

	std::string temp_query_string;
	
	std::map<std::string, std::vector<std::string>>object_values_map;
	std::vector<std::string>object_ids = get_data("objects", "id");

	for(auto id : object_ids){
		std::vector<std::string>object_values;
		std::ostringstream ostring_query;
		ostring_query
			<< "SELECT " << "sprites.path"
			<< " FROM " << "sprites, objects"
			<< " WHERE " << "sprites.id" << " == " << "objects.sprite_id" 
			<< " AND " << "objects.id" << " == " << id
			<< " AND " << "objects.map_id" << " == " << level_id << ";"

			<< "SELECT " << "name"
			<< " FROM " << "objects"
			<< " WHERE " << "map_id" << " == " << level_id 
			<< " AND " << "objects.id" << " == " << id << ";"

			<< "SELECT " << "position_x"
			<< " FROM " << "objects"
			<< " WHERE " << "map_id" << " == " << level_id 
			<< " AND " << "objects.id" << " == " << id << ";"

			<< "SELECT " << "position_y"
			<< " FROM " << "objects"
			<< " WHERE " << "map_id" << " == " << level_id 
			<< " AND " << "objects.id" << " == " << id << ";"

			<< "SELECT " << "sprites.walkable"
			<< " FROM " << "sprites, objects"
			<< " WHERE " << "sprites.id" << " == " << "objects.sprite_id"
			<< " AND " << "objects.map_id" << " == " << level_id 
			<< " AND " << "objects.id" << " == " << id << ";";

		temp_query_string = ostring_query.str();
		object_values = execute_query_with_return(temp_query_string.c_str());
		if (object_values.empty() == false) {
			object_values_map[id] = object_values;
		}

	}

	return object_values_map;

}


std::map<std::string, std::vector<std::string>> sql::get_level_background_value(std::string level_id) {
	
	std::string temp_query_string;
	std::vector<std::string>background_ids = get_data("background", "id");
	std::map<std::string, std::vector<std::string>>object_values_map;
	
	for (auto id : background_ids) {
		std::ostringstream ostring_query;
		std::vector<std::string>background_values;
		ostring_query
			<< "SELECT " << "sprites.path"
			<< " FROM " << "sprites, background"
			<< " WHERE " << "sprites.id" << " == " << "background.sprite_id"
			<< " AND " << "background.map_id" << " == " << level_id 
			<< " AND " << "background.id" << " == " << id << ";"

			<< "SELECT " << "name"
			<< " FROM " << "background"
			<< " WHERE " << "map_id" << " == " << level_id
			<< " AND " << "background.id" << " == " << id << ";"

			<< "SELECT " << "position_x"
			<< " FROM " << "background"
			<< " WHERE " << "map_id" << " == " << level_id
			<< " AND " << "background.id" << " == " << id << ";"

			<< "SELECT " << "position_y"
			<< " FROM " << "background"
			<< " WHERE " << "map_id" << " == " << level_id
			<< " AND " << "background.id" << " == " << id << ";"

			<< "SELECT " << "sprites.walkable"
			<< " FROM " << "sprites, background"
			<< " WHERE " << "sprites.id" << " == " << "background.sprite_id"
			<< " AND " << "background.map_id" << " == " << level_id
			<< " AND " << "background.id" << " == " << id << ";";

		temp_query_string = ostring_query.str();
		object_values_map[id] = execute_query_with_return(temp_query_string.c_str());
		if (background_values.empty() == false) {
			object_values_map[id] = background_values;
		}
	}
	

	return object_values_map;

}


sf::IntRect sql::get_collision_objects(std::vector<std::string> objects) {
	std::vector<std::string>not_walkable = get_data("sprites", "name", "walkable", "0");
	std::vector<sf::IntRect>collision_rects;
	std::string temp_query_string;
	std::ostringstream ostring_query;

	ostring_query
		<< "SELECT " << "size_x"
		<< " FROM " << "sprites"
		<< " WHERE " << "name" << " == '" << objects.at(1) << "';"

		<< "SELECT " << "size_y"
		<< " FROM " << "sprites"
		<< " WHERE " << "name" << " == '" << objects.at(1) << "';";


	temp_query_string = ostring_query.str();
	std::vector<std::string>values = execute_query_with_return(temp_query_string.c_str());

	for (auto indexer : not_walkable) {
		if (indexer == objects.at(1)) {
			std::cout << objects.at(1) << "\t";
			return 
				sf::IntRect(
					std::stoi(objects.at(2)),
					std::stoi(objects.at(3)), 
					std::stoi(values.at(0)),
					std::stoi(values.at(1))
				)
			;
		}
	}
	
	
	


}


std::vector<std::string> sql::get_player_value(std::string player_id) {
	/*
	name
	sprite_id
	map
	level
	pos_x
	pos_y
	faction_id
	*/
	std::string temp_query_string;
	std::ostringstream ostring_query;

	ostring_query
		<< "SELECT " << "name"
		<< " FROM " << "player"
		<< " WHERE " << "id" << " == " << player_id << ";"

		<< "SELECT " << "sprites.path"
		<< " FROM " << "sprites, player"
		<< " WHERE " << "sprites.id" << " == " << "player.sprite_id"
		<< " AND " << "player.id" << " == " << player_id << ";"

		<< "SELECT " << "map"
		<< " FROM " << "player"
		<< " WHERE " << "id" << " == " << player_id << ";"

		<< "SELECT " << "level"
		<< " FROM " << "player"
		<< " WHERE " << "id" << " == " << player_id << ";"

		<< "SELECT " << "position_x"
		<< " FROM " << "player"
		<< " WHERE " << "id" << " == " << player_id << ";"

		<< "SELECT " << "position_y"
		<< " FROM " << "player"
		<< " WHERE " << "id" << " == " << player_id << ";"

		<< "SELECT " << "faction_id"
		<< " FROM " << "player"
		<< " WHERE " << "id" << " == " << player_id << ";";

	temp_query_string = ostring_query.str();

	return execute_query_with_return(temp_query_string.c_str());
}


std::vector<std::string> sql::get_music(std::string map_id) {
	std::ostringstream query;
	query << "SELECT music.path"
		<< " FROM maps, music"
		<< " WHERE maps.id == " << map_id
		<< " AND maps.music_id == music.id;";

	std::string query_string = query.str();

	std::cout << "Buffering music" << std::endl;

	return execute_query_with_return(query_string.c_str());
}



std::vector<std::string> sql::get_quest_text(std::string quest_id, std::string part) {

	std::ostringstream query;
	query << "SELECT text"
		<< " FROM quest_stories"
		<< " WHERE quest_id == " << quest_id
		<< " AND part == " << part << ";"
		<< "SELECT text1"
		<< " FROM quest_stories"
		<< " WHERE quest_id == " << quest_id
		<< " AND part == " << part << ";"
		<< "SELECT text2"
		<< " FROM quest_stories"
		<< " WHERE quest_id == " << quest_id
		<< " AND part == " << part << ";"
		<< "SELECT text3"
		<< " FROM quest_stories"
		<< " WHERE quest_id == " << quest_id
		<< " AND part == " << part << ";"
		<< "SELECT text4"
		<< " FROM quest_stories"
		<< " WHERE quest_id == " << quest_id
		<< " AND part == " << part << ";";

	std::string query_string = query.str();
	return execute_query_with_return(query_string.c_str());
}


std::vector<std::string> sql::get_quest_parts(std::string quest_id) {
	std::ostringstream query;
	query << "SELECT part"
		<< " FROM quest_stories"
		<< " WHERE quest_id == " << quest_id << ";";

	std::string query_string = query.str();
	return execute_query_with_return(query_string.c_str());
}





